src/rust/catloop/runtime.rs

```diff
@@ -0,0 +1,171 @@
+        types::demi_opcode_t,
+    },
+};
+use ::rand::{
+    prelude::SmallRng,
+    SeedableRng,
+};
+use ::std::{
+    collections::HashMap,
+    net::SocketAddrV4,
+};
+
+//======================================================================================================================
+// Structures
+//======================================================================================================================
+
+/// Catloop Runtime
+pub struct CatloopRuntime {
+    /// Ephemeral port allocator.
+    ephemeral_ports: EphemeralPorts,
+    /// Table of queue descriptors, it has one entry for each existing queue descriptor in Catloop LibOS.
+    qtable: IoQueueTable<CatloopQueue>,
+    /// Table for ongoing operations on Catloop LibOS.
+    catloop_qts: HashMap<QToken, (demi_opcode_t, QDesc)>,
+    /// Table for ongoing operations Catmem LibOS.
+    catmem_qts: HashMap<QToken, (demi_opcode_t, QDesc)>,
+}
+
+//==============================================================================
+// Associate Functions
+//==============================================================================
+
+/// Catloop Runtime. This data structure holds all of the cross-queue state for the Catloop libOS.
+impl CatloopRuntime {
+    pub fn new() -> Self {
+        let mut rng: SmallRng = SmallRng::from_entropy();
+        Self {
+            ephemeral_ports: EphemeralPorts::new(&mut rng),
+            qtable: IoQueueTable::<CatloopQueue>::new(),
+            catmem_qts: HashMap::default(),
+            catloop_qts: HashMap::default(),
+        }
+    }
+
+    /// Allocates a new [CatloopQueue] of `qtype`.
+    pub fn alloc_queue(&mut self, qtype: QType) -> QDesc {
+        self.qtable.alloc(CatloopQueue::new(qtype))
+    }
+
+    pub fn free_queue(&mut self, qd: QDesc) {
+        self.qtable.free(&qd);
+    }
+
+    /// Gets the [CatloopQueue] associated with `qd`. If not `qd` does not refer to a valid, then return `EBADF` is returned.
+    pub fn get_queue(&mut self, qd: QDesc) -> Result<&mut CatloopQueue, Fail> {
+        match self.qtable.get_mut(&qd) {
+            Some(queue) => Ok(queue),
+            None => {
+                let cause: String = format!("invalid queue descriptor (qd={:?})", qd);
+                error!("get_queue(): {}", cause);
+                Err(Fail::new(libc::EBADF, &cause))
+            },
+        }
+    }
+
+    /// Inserts a given `qt` queue token into the Catloop queue token table.
+    pub fn insert_catloop_qt(&mut self, qt: QToken, opcode: demi_opcode_t, qd: QDesc) {
+        self.catloop_qts.insert(qt, (opcode, qd));
+    }
+
+    /// Inserts a given `qt` into the Catmem queue token table.
+    pub fn insert_catmem_qt(&mut self, qt: QToken, opcode: demi_opcode_t, qd: QDesc) {
+        self.catmem_qts.insert(qt, (opcode, qd));
+    }
+
+    /// Gets the [QDesc] associated with `qt` in the Catloop queue token table.
+    pub fn get_catloop_qd(&self, qt: QToken) -> Option<&(demi_opcode_t, QDesc)> {
```

Comment:
Could probably be named as: get_catloop_opcode_and_qd?

Also, the comment needs to be corrected.